squint.c

```diff
@@ -0,0 +1,1611 @@
+
+void rename_register1(int *funcBegin, int *funcEnd)
+{
+#define BR 3  /* base register */
+#define MAX_REN_REG 8
+   int *scan;
+   int offset[MAX_REN_REG];
+   int i, numReg = 0;
+
+   /* Abort trivial reg renaming if function calls exist */
+   for (scan = funcBegin; scan <= funcEnd; ++scan)
+      if (*scan == NOP13 && !is_const(scan)) return;
+
+   /* record frame variable in this context */
+   for (scan = funcBegin; scan <= funcEnd; ++scan) {
+      scan = skip_nop(scan, 1);
+
+      if ((*scan & 0x0f2f0000) == 0x050b0000) { // (ldr|str)[b], [fp, #X]
+         int off = (*scan & 0xfff);
+         if  ((*scan & (1<<23)) == 0) off = -off;
+         for (i = 0; i < numReg; ++i) {
+            if (offset[i] == off) break;
+         }
+         if (i == numReg) {
+            offset[numReg++] = off;
+            if (numReg == MAX_REN_REG) break;
+         }
+      }
+   }
+
+   /* discard any frame variables that are not trivial */
+   for (scan = funcBegin; scan <= funcEnd; ++scan) {
+      scan = skip_nop(scan, 1);
+
+      if ((*scan & 0xffffff00) == 0xe28b0000 || // add r0, fp, #X
+          (*scan & 0xffffff00) == 0xe24b0000) { // sub r0, fp, #X
+         int off = *scan & 0xff;
+         if ((*scan & 0xffffff00) == 0xe24b0000) {
+            off = -off;
+         }
+         for (i = 0; i < numReg; ++i) {
+            if (offset[i] == off) break;
+         }
+         if (i != numReg) {
+            --numReg;
+            for(; i < numReg; ++i)
+               offset[i] = offset[i+1];
+         }
+      }
+   }
+
+   if (numReg == 0) return;
+
+   /* create registers for frame vars */
+   for (scan = funcBegin; scan <= funcEnd; ++scan) {
+      scan = skip_nop(scan, 1);
+
+      if ((*scan & 0x0f2f0000) == 0x050b0000) { // (ldr|str)[b], [fp, #X]
+         int off = (*scan & 0xfff);
+         if  ((*scan & (1<<23)) == 0) off = -off;
+         for (i = 0; i < numReg; ++i) {
+            if (offset[i] == off) break;
+         }
+         if (i == numReg) continue; // this frame var not mapped
+         if ((*scan & 0x0f3f0000) == 0x051b0000) { // ldr[b] rX, [fp, #X]
+            *scan = 0xe1a00000 | (*scan & 0x0000f000) | (BR+i);
+         }
+         else { // str[b] rd, [fp, #X] -> mov rX, rd
+            *scan = 0xe1a00000 | ((BR+i) << 12) | ((*scan >> 12) & 0xf);
+         }
+      }
+   }
+
+   /* load frame vars into registers at top of function */
+   for (scan = funcBegin; *scan != NOP; ++scan);
+   for (i = 0; i < numReg; ++i) {
+      *scan++ = 0xe51b0000 | ((BR + i) << 12) |
+                ((offset[i] < 0) ? -offset[i] : (offset[i] | (1<<23)));
+   }
+}
+
+/**********************************************************/
+/********* Peephole optimization driver function **********/
+/**********************************************************/
+
+int squint_opt(int *begin, int *end)
+{
+   int optApplied = 0 ;
+   int *scan = begin;
+   int *tmpbuf = (int *) malloc((end-begin)*sizeof(int));
```

Comment:
I think you will need `(end - begin + 1)`, where `+ 1` is for the extra sentinel added on Line 567 (or even `+ 2` if you also want to add sentinel at the beginning).

Consider this example: there is only one function in the input executable, `end - begin` will be the number of instructions in the function (without sentinel).